File: ruby/first_non_repeating_char.rb
# frozen_string_literal: true

def first_non_repeating_char(string)
  tally = string.chars.tally
  string.each_char { |c| return c if tally[c] == 1 }

  'No non-repeating characters'
end

p first_non_repeating_char('swiss') # => 'w'
p first_non_repeating_char('apple') # => 'a'
p first_non_repeating_char('') # => 'No non-repeating characters'
p first_non_repeating_char('mama') # => 'No non-repeating characters'

File: typescript/firstNonRepeatingChar.ts
function firstNonRepeatingChar(str) {
  let tally = {};

  for (let char of str) {
    // if (tally[char]) {
    //   tally[char]++;
    // } else {
    //   tally[char] = 1;
    // }

    // OR (Short circuiting)
    tally[char] = (tally[char] || 0) + 1;
  }

  for (let char of str) {
    if (tally[char] === 1) {
      return char;
    }
  }
}

console.log(firstNonRepeatingChar("abacddbec"));
console.log(firstNonRepeatingChar("swiss"));
console.log(firstNonRepeatingChar("mummyuy"));
